src/libre/re_analysis.c

```diff
@@ -0,0 +1,77 @@
+#include "re_analysis.h"
+
+#include <string.h>
+
+
+struct analysis_env {
+	unsigned group_id;
+};
+
+static void
+analysis_iter(struct analysis_env *env, struct ast_expr *n);
+
+static void
+analysis_char_class(struct analysis_env *env, struct re_char_class_ast *cca);	
+
+void
+re_ast_analysis(struct ast_re *ast)
+{
+	struct analysis_env env;
+	if (ast == NULL) { return; }
+	memset(&env, 0x00, sizeof(env));
+
+	assert(ast->expr != NULL);
+	analysis_iter(&env, ast->expr);
+}
+
+/* Now that the AST is complete, do a pass filling in details. */
+/* TODO: extract traversal from the callback, so the individual
+ * analysis passes have an obvious focus. */
+static void
+analysis_iter(struct analysis_env *env, struct ast_expr *n)
+{
+	switch (n->t) {
+	case AST_EXPR_EMPTY:
+	case AST_EXPR_LITERAL:
+		break;
+	case AST_EXPR_CONCAT:
+		analysis_iter(env, n->u.concat.l);
+		analysis_iter(env, n->u.concat.r);
+		break;
+	case AST_EXPR_ALT:
+		analysis_iter(env, n->u.alt.l);
+		analysis_iter(env, n->u.alt.r);
+		break;
+
+	case AST_EXPR_ANY:
+	case AST_EXPR_MANY:
+	case AST_EXPR_FLAGS:
+		break;
+
+	case AST_EXPR_REPEATED:
+		analysis_iter(env, n->u.repeated.e);
+		break;
+	case AST_EXPR_CHAR_CLASS:
+		analysis_char_class(env, n->u.char_class.cca);
+		break;
+	case AST_EXPR_GROUP:
+		/* assign group ID */
+		env->group_id++;
+		n->u.group.id = env->group_id;
+		analysis_iter(env, n->u.group.e);
+		break;
```

Comment:
There are some things that the `re_ast_analysis()` pass is doing which could definitely be done on-the-fly during parse, but i know there's more coming for anchors there, and i'm happy to leave that as-is and think about it again after things settle down.

For example, i think group numbering would be pretty natural to supply with a SID action keeping the current maximum in `struct act_state` and providing a sequential ID each time it's called.

Reply:
The group anchoring could definitely move into the parser, but it was a simple case to prove out the top-down analysis pass.

As far as I can tell, a separate analysis pass is either necessary to make anchoring work properly, or will greatly simplify the implementation.